src/ucp/rma/amo_sw.c

```diff
@@ -194,39 +194,182 @@ DEFINE_AMO_SW_OP(64)
+{
+    ucp_ep_h reply_ep;
+    ucp_request_t *req;
+    ucs_status_t status;
+
+    reply_ep = ucp_worker_get_ep_by_ptr(worker, flush_ctx->req_hdr.req.ep_ptr);
+
+    if (flush_ctx->req_hdr.req.reqptr == 0) {
+        ucp_rma_sw_send_cmpl(reply_ep);
+    } else {
+        req = ucp_request_get(worker);
+        if (req == NULL) {
+            ucs_error("failed to allocate atomic reply");
+        } else {
+            req->send.ep               = reply_ep;
+            req->send.atomic_reply.req = flush_ctx->req_hdr.req.reqptr;
+            req->send.length           = flush_ctx->req_hdr.length;
+            req->send.uct.func         = ucp_progress_atomic_reply;
+            ucp_request_send(req, 0);
+        }
+    }
+
+    ucp_rkey_destroy(flush_ctx->rkey);
+    status = ucp_mem_unmap(worker->context, flush_ctx->memh);
+    if (status != UCS_OK) {
+        ucs_error("failed to unmap memory handle %p with error %s",
+                  flush_ctx->memh, ucs_status_string(status));
+    }
+
+    ucs_free(flush_ctx);
+}
+
+static void ucp_amo_sw_loopback_flush_cb(ucp_request_t *req)
+{
+    ucp_atomic_flush_ctx_t *flush_ctx = req->send.flush_ctx;
+    ucp_worker_h worker               = req->send.ep->worker;
+
+    ucp_amo_sw_loopback_send_reply(worker, flush_ctx);
+    ucp_request_put(req);
+}
+
+static ucs_status_t
+ucp_amo_sw_loopback_post(ucp_worker_h worker, ucp_atomic_req_hdr_t *amo_req_hdr)
+{
+    ucp_atomic_flush_ctx_t *flush_ctx;
+    ucp_mem_map_params_t params;
+    void *rkey_buffer;
+    size_t rkey_buffer_size;
+    void *request;
+    ucs_status_t status, status_unmap;
+
+    /* TODO: optimization: can use `am_flags & UCT_CB_PARAM_FLAG_DESC` to avoid
+     *       extra allocation and copy for some transports */
+    flush_ctx = ucs_malloc(sizeof(*flush_ctx), "loop back amo flush ctx");
+    if (flush_ctx == NULL) {
+        ucs_error("failed to allocate loopback AMO flush context");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    flush_ctx->req_hdr = *amo_req_hdr;
+    flush_ctx->value   = *(uint64_t *)(amo_req_hdr + 1);
+
+    params.field_mask  = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
+                         UCP_MEM_MAP_PARAM_FIELD_LENGTH  |
+                         UCP_MEM_MAP_PARAM_FIELD_FLAGS;
+    params.address     = (void *)amo_req_hdr->address;
+    params.length      = amo_req_hdr->length;
+    params.flags       = 0;
+
+    /*
+     * TODO: rkey can be passed as a part of AM to avoid
+     *       mem_map/rkey_pack/rkey_pack.
+     * NOTE: if rcache is enabled then registration should be "for free" since
+     *       destination memory has been registered.
+     */
+    status = ucp_mem_map(worker->context, &params, &flush_ctx->memh);
+    if (status != UCS_OK) {
+        ucs_error("failed to map memory region %p length %zu status %s",
+                  params.address, params.length, ucs_status_string(status));
+        goto err_free_ctx;
+    }
+
```

Comment:
can we detect such situation on sender side and add packed rkey to sw amo request?

Reply:
I'm not sure that we can pack again unpacked address on initiator side.
another proposal - maybe better to store rkey to memh (if it's not yet) and pass target memh pointer as part of key, then we can send only this pointer in amo request, WDYT?

Reply:
ok. but we must do it only when needed (to not increase rkey size) and cache the unpacked rkey on target memh so 2nd 3rd,... operations will not have to pack/unpack rkey again